libyul/Object.cpp

```diff
@@ -65,9 +65,45 @@ set<YulString> Object::dataNames() const
 {
 	set<YulString> names;
 	names.insert(name);
-	for (auto const& subObject: subIndexByName)
-		names.insert(subObject.first);
-	// The empty name is not valid
+	for (auto const& subObjectNode: subObjects)
+	{
+		names.insert(subObjectNode->name);
+		if (auto const* subObject = dynamic_cast<Object const*>(subObjectNode.get()))
+			for (YulString const& subSubObj: subObject->dataNames())
+				names.insert(YulString{subObject->name.str() + "." + subSubObj.str()});
+	}
+
 	names.erase(YulString{});
 	return names;
 }
+
+map<YulString, shared_ptr<ObjectNode>> Object::subObjectsByDataName() const
+{
+	map<YulString, shared_ptr<ObjectNode>> objectsByDataName;
+	for (shared_ptr<ObjectNode> const& subObjectNode: subObjects)
+		if (auto const* subObject = dynamic_cast<Object const*>(subObjectNode.get()))
+		{
+			set<YulString> subObjDataNames = subObject->dataNames();
+			for (YulString const& subObjDataName: subObjDataNames)
```

Comment:
With this many nesting levels names get confusing if the `sub` and `subSub` prefixes are not used consistently.

I'd also prefix with `qualified` in places where names are actually "paths" with multiple components.

```suggestion
			for (YulString const& qualifiedSubSubObjectName: subObjDataNames)
```